File: src/main/java/uk/co/eelpieconsulting/feedlistener/model/Subscription.java
package uk.co.eelpieconsulting.feedlistener.model;

import java.util.Date;

import org.bson.types.ObjectId;
import org.mongodb.morphia.annotations.Entity;
import org.mongodb.morphia.annotations.Id;
import org.mongodb.morphia.annotations.Indexed;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;

@Entity("subscriptions")
@JsonPropertyOrder({"id", "name", "channel", "url"})
public abstract class Subscription {
	
    @Id
    ObjectId objectId;
    
    private String id, name, username;  
    private Date lastRead, latestItemDate;
    private String error;
  
    @Indexed
    private String channelId;
        
	public final String getId() {
		return id;
	}

	public final void setId(String id) {
		this.id = id;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public Date getLastRead() {
		return lastRead;
	}

	public void setLastRead(Date lastRead) {
		this.lastRead = lastRead;
	}

	public Date getLatestItemDate() {
		return latestItemDate;
	}

	public void setLatestItemDate(Date latestItemDate) {
		this.latestItemDate = latestItemDate;
	}
	
	public String getChannelId() {
		return channelId;
	}

	public void setChannelId(String channelId) {
		this.channelId = channelId;
	}

	public String getUsername() {
		return username;
	}

	public void setUsername(String username) {
		this.username = username;
	}

	public String getError() {
		return error;
	}

	public void setError(String error) {
		this.error = error;
	}

	@Override
	public String toString() {
		return "Subscription [channelId=" + channelId + ", error=" + error
				+ ", id=" + id + ", lastRead=" + lastRead + ", latestItemDate="
				+ latestItemDate + ", name=" + name + ", username=" + username
				+ "]";
	}
	
}

File: src/main/kotlin/uk/co/eelpieconsulting/feedlistener/model/Subscription.kt
package uk.co.eelpieconsulting.feedlistener.model

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonPropertyOrder
import dev.morphia.annotations.Entity
import dev.morphia.annotations.Id
import dev.morphia.annotations.Indexed
import org.bson.types.ObjectId
import uk.co.eelpieconsulting.feedlistener.rss.classification.FeedStatus
import java.util.*

@Entity("subscriptions")
@JsonPropertyOrder("id", "name", "channel", "url")
@JsonInclude(JsonInclude.Include.NON_NULL)
abstract class Subscription {
    @Id
    var objectId: ObjectId? = null

    lateinit var id: String
    var name: String? = null
    lateinit var username: String
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ssXXX")
    var lastRead: Date? = null
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ssXXX")
    var latestItemDate: Date? = null
    var error: String? = null
    var etag: String? = null
    var httpStatus: Int? = null
    var itemCount= 0L
    var lastModified: Date? = null
    var classifications: Set<FeedStatus>? = emptySet()

    @Indexed
    lateinit var channelId: String

}